lib/vrl/DESIGN.md

```diff
@@ -0,0 +1,219 @@
+    * [Fallibility](#fallibility)
+    * [Signatures](#signatures)
+  * [Errors](#errors)
+* [Patterns](#patterns)
+  * [Error Chaining](#error-chaining)
+
+## The Zen of VRL
+
+In the spirit of [The Zen of Python][PEP20].
+
+- Beautiful is better than ugly.
+- Explicit is better than implicit.
+- Simple is better than complex.
+- Sparse is better than dense.
+- Readability trumps writing convenience.
+- Special cases aren't special enough to break the rules.
+- Although practicality beats purity.
+- Function calls over syntax.
+- Errors may never occur at run-time.
+- Unless explicitly marked as such.
+- There should be one — and preferably only one — obvious way to do it.
+- Performance matters.
+- Provide observability remapping solutions, nothing else.
+
+[PEP20]: https://www.python.org/dev/peps/pep-0020/
+
+## Why VRL
+
+VRL exists to solve the need for a _simple_ and _performant_ **domain specific
+language** (DSL) to remap observability data.
+
+Its purpose is to hit a sweet spot between Turing complete languages such as
+Lua, and static transforms such as Vector's `rename_fields`. It should be
+flexible enough to cover most remapping use-cases, without requiring operators
+to write complex scripts that are difficult to reason about and incur
+a significant performance penalty.
+
+See the [introduction blog post][blog] for more details on the **why**.
+
+[blog]: https://vector.dev/blog/vector-remap-language/#preamble
+
+## Target Audience
+
+VRL has a specialized purpose of solving observability data remapping needs.
+Because of this purpose, the language is mostly used by people who manage
+infrastructure at their organisations. The role of this group is usually
+referred to as "operator", "devops" (Development and Operations) or "SRE" (Site
+Reliability Engineer).
+
+One common generalization of this group is that they are focused on maintaining
+infrastructure within an organization, and often write and maintain their own
+software to achieve this goal. They are adept enough at programming to achieve
+their goals, but have no need or desire to be as skilled in programming as
+dedicated software engineers, because their time is best spent elsewhere.
+
+As with everything, there are exceptions to the rule, and many people in this
+group _are_ highly skilled software engineers, but VRL must capture the largest
+possible segment of this group, and should therefor be limited in complexity.
+
+Therefor, when extending the feature set of VRL, design **the feature for the
+intended target audience**, which will likely mean choosing different trade-offs
+than you'd make if you were to design the feature for your personal needs.
+
+## Language Limits
+
+There are a number of features that we've so far rejected to implement in the
+language. This might change in the future, but there should be a good reason to
+do so.
+
+- modules (see: [#5507][]
+- classes
+- user-defined functions
+- `goto` statements
+- network calls (see [#4517][])
```

Comment:
Can we provide short justifications for each? They should connect back to our principles.